### PR TITLE
[LOG4J2-2803] Create standardized dependency injection API

### PR DESCRIPTION
**Pull request by jvz**
This replaces the previous attempt at unifying dependency injection and
configuration from a CDI-like API to an API more like javax.inject used
by libraries like Feather and Guice.

https://cwiki.apache.org/confluence/display/LOGGING/Dependency+Injection+and+Configuration

### CODE REVIEW

File: log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/ConfigurationInjector.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
  * @param <Ann> plugin annotation this injector uses
  * @param <Cfg> configuration class
  */
+@Deprecated
```
Comment (jvz) [inline log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/ConfigurationInjector.java:38]: To be deleted.

File: log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/InjectorStrategy.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
 @Documented
 @Retention(RetentionPolicy.RUNTIME)
 @Target(ElementType.ANNOTATION_TYPE)
-// TODO: annotation processor to validate type matches (help avoid runtime errors)
+@Deprecated
```
Comment (jvz) [inline log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/InjectorStrategy.java:30]: To be deleted.

File: log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/PluginAttributeParser.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            Map.entry(char.class, PluginAttribute::defaultChar),
+            Map.entry(Character.class, PluginAttribute::defaultChar),
+            Map.entry(short.class, PluginAttribute::defaultShort),
+            Map.entry(Short.class, PluginAttribute::defaultShort),
+            Map.entry(Class.class, PluginAttribute::defaultClass)
+    );
+
+    private final Function<String, String> stringSubstitutionStrategy;
+
+    @Inject
```
Comment (jvz) [inline log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/inject/PluginAttributeParser.java:57]: Note how the parsing strategies for nodes can themselves be dependency-injected. In this case, we're using a `StrSubstitutor` function.

File: log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/processor/PluginEntry.java [hunk 4]
```diff
@@ -17,14 +17,10 @@
 
 package org.apache.logging.log4j.plugins.processor;
 
-import java.io.Serializable;
-
 /**
  * Memento object for storing a plugin entry to a cache file.
  */
-public class PluginEntry implements Serializable {
```
Comment (jvz) [inline log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/processor/PluginEntry.java:25]: https://issues.apache.org/jira/browse/LOG4J2-3228 for the full story

**Review by jvz — COMMENTED** [review]
Left some line comments for context.

**Comment by gubasza** [comment]
>

**Review by gubasza — APPROVED** [review]
gubasza

**Comment by jvz** [comment]
@rgoers you can have a look now

File: log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/di/ReflectiveCallerContext.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import java.lang.reflect.InaccessibleObjectException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Represents the calling context {@link Injector} uses for reflection operations. This allows for customizing the caller class
+ * for invocations of {@link AccessibleObject#setAccessible(boolean)}.
+ */
+@FunctionalInterface
+public interface ReflectiveCallerContext {
```
Comment (jvz) [inline log4j-plugins/src/main/java/org/apache/logging/log4j/plugins/di/ReflectiveCallerContext.java:32]: - [ ] Consider using `MethodHandles.Lookup` instead

**Comment by jvz** [comment]
I'd like to merge this sometime within the next couple weeks unless there are some issues raised.

**Pull request merged**